mitodl/mitxonline
# Updates flexible pricing to add a unique constraint on submissions

#### Pre-Flight checklist


#### What are the relevant tickets?

Fixes #803 
Fixes #804 

#### What's this PR do?

Adds a unique constraint on the `user` and `courseware_object` fields (or, rather, the GenericRelation fields it uses) to the `FlexiblePrice` object. Additionally updates the logic for the `FlexiblePriceRequestForm` to check for an existing request:
- If there's an existing request not in the Review state, submitting will fail. (The user should never get to this point but they will see a 500 if they do.) 
- The logic to determine the last submission from the user and the related courseware object now takes into account the set courseware object for the form. If the object is a standalone course or a program, it will use that. If it's a course in a program, it will return the program it belongs to. (You cannot submit a flexible price request, nor create a form for, an individual course in a program.) 

#### How should this be manually tested?

Automated tests should pass. Manual testing will require a set of courses created, ideally one independent of a program and two that are in the same program.

1. Create Flexible Price Request Forms under each course. The ones for the program courses can be associated with either the course _or_ the program. (One of each is a good test.)
2. Attempt to submit a flexible price request using one of the program course forms. The request should be successful.
3. Attempt to submit a flexible price request using the other program course form. You should not see the form and should instead see the status page (as configured in the form) for the request you entered. 
4. Attempt to submit a flexible price request using the independent course form. The request should be successful. 

You can also use a different user and run the same test as above, except with steps 3 and 4 reversed, if you've assigned one form to a course and one form to a program. In either case, the generated `FlexiblePrice` record should be for the program.
